From 66ef2f3dfcb7cee2e3a6cac16ff3272a150b59e2 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Sun, 04 Oct 2026 10:13:04 +0200
Subject: [PATCH] Revert "Fix Use-After-Free in NXP NFC HAL timer teardown"

Sadly this AI generated fix for AI generated bug doesn't work properly
and every once in a while will just get stuck on something, logspam and
drain your battery...

This reverts commit f032114b8752cf3977854d87b1acc268a0637749.

Change-Id: I19e2f5d701cb5c8fb7c2047b9b402696bf52457e
---

diff --git a/pn8x/halimpl/hal/phNxpNciHal.cc b/pn8x/halimpl/hal/phNxpNciHal.cc
index e3c39d9..a246e52 100644
--- a/pn8x/halimpl/hal/phNxpNciHal.cc
+++ b/pn8x/halimpl/hal/phNxpNciHal.cc
@@ -385,11 +385,8 @@
   if (status != NFCSTATUS_SUCCESS) {
     if (NFCSTATUS_SUCCESS != phNxpNciHal_fw_mw_ver_check()) {
       NXPLOG_NCIHAL_D("Chip Version Middleware Version mismatch!!!!");
-      intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-      nxpncihal_ctrl.gDrvCfg.nClientId = 0;
       phOsalNfc_Timer_Cleanup();
       phTmlNfc_Shutdown_CleanUp();
-      phDal4Nfc_msgrelease(client_id);
       status = NFCSTATUS_FAILED;
     } else {
       NXPLOG_NCIHAL_E("FW download failed, Continue NFCC init");
@@ -2069,11 +2066,8 @@
           (pphTmlNfc_TransactCompletionCb_t)&phNxpNciHal_read_complete, NULL);
       if (status != NFCSTATUS_PENDING) {
         NXPLOG_NCIHAL_E("TML Read status error status = %x", status);
-        intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-        nxpncihal_ctrl.gDrvCfg.nClientId = 0;
         phOsalNfc_Timer_Cleanup();
         phTmlNfc_Shutdown();
-        phDal4Nfc_msgrelease(client_id);
         status = NFCSTATUS_FAILED;
       }
       break;
@@ -2191,8 +2185,6 @@
     status = phTmlNfc_ReadAbort();
     status = phTmlNfc_WriteAbort();
 
-    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2203,7 +2195,7 @@
 
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(client_id);
+    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
 
@@ -2243,8 +2235,6 @@
     status = phTmlNfc_ReadAbort();
     status = phTmlNfc_WriteAbort();
 
-    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2255,7 +2245,7 @@
 
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(client_id);
+    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
 
diff --git a/snxxx/halimpl/hal/phNxpNciHal.cc b/snxxx/halimpl/hal/phNxpNciHal.cc
index 6e816cf..2e0edd8 100644
--- a/snxxx/halimpl/hal/phNxpNciHal.cc
+++ b/snxxx/halimpl/hal/phNxpNciHal.cc
@@ -490,12 +490,9 @@
       fw_download_success = TRUE;
     } else if (status == NFCSTATUS_FW_CHECK_INTEGRITY_FAILED ||
                (phNxpNciHal_fw_mw_ver_check() != NFCSTATUS_SUCCESS)) {
-      intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-      nxpncihal_ctrl.gDrvCfg.nClientId = 0;
       phOsalNfc_Timer_Cleanup();
       phNxpTempMgr::GetInstance().Reset();
       phTmlNfc_Shutdown_CleanUp();
-      phDal4Nfc_msgrelease(client_id);
       return NFCSTATUS_CMD_ABORTED;
     }
 
@@ -2570,8 +2567,6 @@
     status = phTmlNfc_ReadAbort();
     status = phTmlNfc_WriteAbort();
 
-    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2583,7 +2578,7 @@
     phNxpTempMgr::GetInstance().Reset();
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(client_id);
+    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
 
@@ -2653,8 +2648,6 @@
     if (status != NFCSTATUS_SUCCESS) {
       NXPLOG_TML_E("phTmlNfc_ReadAbort Failed");
     }
-    intptr_t client_id = nxpncihal_ctrl.gDrvCfg.nClientId;
-    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
@@ -2670,7 +2663,7 @@
     phNxpTempMgr::GetInstance().Reset();
     phTmlNfc_CleanUp();
 
-    phDal4Nfc_msgrelease(client_id);
+    phDal4Nfc_msgrelease(nxpncihal_ctrl.gDrvCfg.nClientId);
 
     memset(&nxpncihal_ctrl, 0x00, sizeof(nxpncihal_ctrl));
   }
diff --git a/snxxx/halimpl_v2/hal/phNxpNciHal.cc b/snxxx/halimpl_v2/hal/phNxpNciHal.cc
index afe6e88..dd965b7 100644
--- a/snxxx/halimpl_v2/hal/phNxpNciHal.cc
+++ b/snxxx/halimpl_v2/hal/phNxpNciHal.cc
@@ -321,7 +321,6 @@
       fw_download_success = true;
     } else if (status == NFCSTATUS_FW_CHECK_INTEGRITY_FAILED ||
                (phNxpNciHal_fw_mw_ver_check() != NFCSTATUS_SUCCESS)) {
-      nxpncihal_ctrl.gDrvCfg.nClientId = 0;
       phOsalNfc_Timer_Cleanup();
       phNxpTempMgr::GetInstance().Reset();
       phTmlNfc_Shutdown_CleanUp();
@@ -2305,7 +2304,6 @@
     phNxpNciHal_complete(NFCSTATUS_SUCCESS, PHNXP_NCIHAL_OP_CLOSE);
     /* Abort any pending read and write */
     (void)phTmlNfc_ReadAbort();
-    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     (void)phTmlNfc_Shutdown();
@@ -2367,7 +2365,6 @@
     if (status != NFCSTATUS_SUCCESS) {
       NXPLOG_TML_E("phTmlNfc_ReadAbort Failed");
     }
-    nxpncihal_ctrl.gDrvCfg.nClientId = 0;
     phOsalNfc_Timer_Cleanup();
 
     status = phTmlNfc_Shutdown();
